spyder_kernels/customize/spydercustomize.py

```diff
@@ -862,46 +917,92 @@ def runcell(cellname, filename):
         # AttributeError --> systematically raised in Python 3
         pass
     ipython_shell = get_ipython()
+    try:
+        # Get code from spyder
+        cell_code = _frontend_request().runcell(cellname, filename)
+        if not filename:
+            filename = _frontend_request().current_filename()
+    except (CommError, TimeoutError):
+        if hasattr(ipython_shell, "cell_code"):
+            # This is the old way of transmitting informations
+            cell_code = ipython_shell.cell_code
+            del ipython_shell.cell_code
+        else:
+            _print("--Run Cell Error--\n"
+                   "Please use only with Spyder; ")
+            return
+
     namespace = _get_globals()
+    # Save previous __file__
+    namespace_has_file = '__file__' in namespace
+    if namespace_has_file:
+        namespace_filename = namespace['__file__']
     namespace['__file__'] = filename
-    try:
-        cell_code = ipython_shell.cell_code
-    except AttributeError:
-        _print("--Run Cell Error--\n"
-               "Please use only through Spyder's Editor; "
-               "shouldn't be called manually from the console")
-        return
 
     # Trigger `post_execute` to exit the additional pre-execution.
     # See Spyder PR #7310.
     ipython_shell.events.trigger('post_execute')
 
-    ipython_shell.run_cell(cell_code)
-    namespace.pop('__file__')
-    del ipython_shell.cell_code
+    try:
+        exec(compile(cell_code, filename, 'exec'), namespace)
```

Comment:
I don't like this. I think we should run `ipython_shell.run_cell`, as before.

Reply:
Also in #134.